---
title: "Setup Shiny Assets"
author: "Shackett"
date: "`r format(Sys.time(), '%d %B, %Y')`"
format:
  html:
    embed-resources: true
    toc: true
    theme: minty
    highlight: tango
    code-block-bg: true
    code-block-border-left: "#5BB867"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Organize Dependencies

```{r globals}
params <- list(
  project_path = Sys.getenv("doomics_zfs_root"),
  repo_path = Sys.getenv("doomics_repo_path"),
  do_drive_token = Sys.getenv("doomics_drive_working_uri"),
  shock_token = Sys.getenv("doomics_shock_token"),
  out_base_dir = Sys.getenv("doomics_out_base_dir")
)

calibase::configure_google_access()

source(file.path(params$repo_path, "bioinformatics", "utils", "data_prep.R"))
source(file.path(params$repo_path, "bioinformatics", "3_interactive", "results_browser", "interactive.R"))

suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(ggplot2))
```

# Load Results

```{r load_regression_outputs}
diffex_outdir <- file.path(params$project_path, params$out_base_dir, "differential_abundances")

lm_models <- readRDS(file = file.path(diffex_outdir, "out_files", "lm_models.Rds"))
lm_params <- readRDS(file = file.path(diffex_outdir, "out_files", "target_model_parameters.Rds"))
model_signif_raw <- readRDS(file = file.path(diffex_outdir, "out_files", "model_significance.Rds"))

model_summaries <- lm_models %>%
  tidyr::unnest(model_summary)
```

```{r load_data}
wide_age_based_phenotypes <- load_wide_age_based_phenotypes(params$shock_token)

feature_design_list <- construct_features_with_design(params, met_input = "valid") 
# apply feature filters applied during regression modeling
# basically removing a handful of identical features
feature_design_list$features_with_design <- feature_design_list$features_with_design %>%
  semi_join(model_signif_raw, by = c("data_type", "groupId"))

# write each value of list to variables
features_with_design <- feature_design_list$features_with_design %>%
  # add additional sample features
  dplyr::left_join(wide_age_based_phenotypes, by = c("Mouse.ID", Age = "blood_draw_age_coarse", "Generation")) %>%
  dplyr::select(-sampleId)

feature_attributes <- dplyr::bind_rows(
  feature_design_list$proteomics_features,
  feature_design_list$small_molecules_features,
  feature_design_list$peptidomics_features %>%
    dplyr::mutate(data_type = "peptides",
                  NameUnique = PeptideSequence),
  feature_design_list$olink_features %>%
    dplyr::mutate(data_type = "olink",
                  NameUnique = Assay,
                  GeneName = NameUnique
                  )
  ) %>%
  dplyr::mutate(feature_name = dplyr::case_when(!is.na(compoundName) ~ compoundName,
                                                !is.na(NameUnique) ~ NameUnique)) %>%
  dplyr::select(data_type, groupId, feature_name, gene = GeneName, is_unknown) %>%
  dplyr::mutate(is_unknown = ifelse(is.na(is_unknown), FALSE, is_unknown)) %>%
  dplyr::group_by(feature_name) %>%
  dplyr::mutate(entry = 1:n()) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(
    feature_name = dplyr::case_when(
      entry == 1 ~ feature_name,
      TRUE ~ as.character(glue::glue("{feature_name} ({entry})"))
    )
  ) %>%
  dplyr::select(-entry)

# check that features and attributes match
missing_attributes <- features_with_design %>%
  dplyr::distinct(data_type, groupId) %>%
  dplyr::anti_join(feature_attributes, by = c("data_type", "groupId"))

if (nrow(missing_attributes) != 0) {
  missing_attributes %>%
  glue::glue_data("No attributes for {data_type}, groupId: {groupId}. ") %>%
  stop()
}

unmatched_features <- features_with_design %>%
  dplyr::distinct(data_type, groupId) %>%
  dplyr::anti_join(feature_attributes, by = c("data_type", "groupId"))
if (nrow(unmatched_features) != 0) {
  stop("Some features did not have matching attributes")
}
```

## Reorganize Results and Validate that Different Files are Consistent

```{r read_other_summaries}
model_signif <- model_signif_raw %>%
  dplyr::mutate(plotting_priority = dplyr::case_when(
    stringr::str_detect(groupName, "unk") ~ "unknown",
    TRUE ~ "identified feature"),
    plotting_priority = factor(plotting_priority, levels = c("identified feature", "unknown"))) %>%
  dplyr::inner_join(
    feature_attributes %>% dplyr::select(data_type, groupId, feature_name),
    by = c("data_type", "groupId")
  )

features_with_design <- features_with_design %>%
  # add data modality
  dplyr::inner_join(
    model_signif %>%
      dplyr::distinct(data_modality, data_type),
    by = "data_type"
    ) %>%
  dplyr::relocate(data_modality, .after = dataset)

check_regression_summaries(
  features_with_design,
  lm_models,
  lm_params,
  model_signif,
  model_summaries
)
```

## Project Out Batch Effects

```{r correcting_batch_effects}
# project out nuisance variation and cluster features

# pull out estimates of cross-sectional (xs) batch effects
# summarize early G8 and LBD batch effects for each feature
batch_adjustments <- lm_models %>%
  dplyr::filter(model_name == "cross-sectional lm (age + lifespan), no DDM") %>%
  dplyr::select(-fits, -model_name) %>%
  tidyr::unnest(params) %>%
  dplyr::filter(term %in% c("early G8", "late blood draw")) %>%
  dplyr::select(data_type, groupId, term, estimate) %>%
  tidyr::spread(term, estimate) %>%
  dplyr::rename(EG8 = `early G8`, LBD = `late blood draw`) %>%
  dplyr::mutate(Ref = 0) %>%
  tidyr::gather(xs_batch, xs_batch_adj, -data_type, -groupId) %>%
  dplyr::mutate(xs_batch = factor(xs_batch, levels = levels(features_with_design$xs_batch)))

projected_features_with_design <- features_with_design %>%
  # add feature attributes
  dplyr::left_join(feature_attributes, by = c("data_type", "groupId")) %>%
  # add batch ajustments
  dplyr::left_join(batch_adjustments, by = c("data_type", "groupId", "xs_batch")) %>%
  dplyr::mutate(corrected_log2_abundance = normalized_log2_abundance - xs_batch_adj) %>%
  # recalculate initial abundances and fold-changes
  dplyr::group_by(dataset, groupId, Mouse.ID) %>%
  dplyr::mutate(corrected_initial_abundance = ifelse("8 months" %in% Age, corrected_log2_abundance[Age == "8 months"], NA_real_)) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(corrected_log2_fold_change = corrected_log2_abundance - corrected_initial_abundance) %>%
  dplyr::select(-xs_batch_adj)
```

## Cluster Features Using Gaussian Mixture Models

```{r clustering}
# also remove sex effects then cluster features into K clusters

sex_effects <- lm_models %>%
  dplyr::filter(model_name == "cross-sectional lm (age + lifespan), no DDM") %>%
  dplyr::select(-fits, -model_name) %>%
  tidyr::unnest(params) %>%
  dplyr::filter(term == "sex") %>%
  dplyr::select(data_type, groupId, sex_effect = estimate)

sex_controlled_projection <- projected_features_with_design %>%
  dplyr::left_join(sex_effects, by = c("data_type", "groupId")) %>%
  dplyr::mutate(sex_controlled_abundance = corrected_log2_abundance - as.integer(Sex)*sex_effect) %>%
  # drop DDM
  dplyr::filter(!is_ddm_sample) %>%
  dplyr::filter(data_modality %in% c("metabolomics", "lipidomics", "proteomics")) %>%
  # center each feature
  dplyr::group_by(feature_name) %>%
  dplyr::mutate(sex_controlled_abundance = sex_controlled_abundance - mean(sex_controlled_abundance, na.rm = TRUE)) %>%
  dplyr::ungroup()
  
# k-means clustering with Gaussian Mixture Models

gmm_df <- sex_controlled_projection %>%
  dplyr::select(
    sample = feature_name, # we're clustering features
    feature = unique_sample_id,
    observation = sex_controlled_abundance
    )

gmm_clusters_cache_path <- file.path(Sys.getenv("doomics_scratch_dir"), "gmm_clusters.Rds")
if (file.exists(gmm_clusters_cache_path)) {
  gmm_clusters <- readRDS(gmm_clusters_cache_path)
} else {
  
  gmm_clusters <- calicostats::k_clustering(
    gmm_df,
    k_vals = seq(2, 30, by = 2),
    method = "gaussian_mixture_EM",
    n_initializations = 5L,
    return_plot = TRUE,
    noise_type = "constant"
    )

  saveRDS(gmm_clusters, gmm_clusters_cache_path)
}

best_clusters <- gmm_clusters$cluster_fit %>%
  dplyr::filter(k == 20) %>%
  dplyr::select(-logLik) %>%
  tidyr::unnest_wider(cluster)

feature_clusters <- best_clusters$membership[[1]] %>%
  dplyr::arrange(desc(responsibility)) %>%
  dplyr::group_by(sample) %>%
  dplyr::slice(1) %>%
  dplyr::ungroup()  %>%
  dplyr::select(feature_name = sample, cluster) %>%
  dplyr::mutate(cluster = factor(cluster, levels = sort(unique(cluster))))
```

## Organize Results as T\*Omic Objects To Take Advantage of *romic*

```{r romic_formatting}
tidy_features <- projected_features_with_design %>%
  dplyr::semi_join(features_with_design %>%
                     dplyr::count(data_type, groupId) %>%
                     dplyr::filter(n >= (max(n)/4)),
                   by = c("data_type", "groupId")) %>%
  # add clusters
  dplyr::left_join(feature_clusters, by = "feature_name") %>%
  dplyr::mutate(
    cluster = ifelse(is.na(cluster), -1, cluster),
    cluster = factor(cluster, levels = sort(unique(cluster)))
    ) %>%
  # add general features
  romic::create_tidy_omic(feature_pk = "feature_name",
                          feature_vars = c("data_type", "data_modality", "groupId", "gene", "is_unknown", "cluster"),
                          sample_pk = "unique_sample_id",
                          sample_vars = c("set", "sample", "Mouse.ID", "blood_draw_age_actual",
                                          "Age", "Sex", "lifespan", "lifespan_remaining", "fraction_of_life_lived", "is_ddm_sample",
                                          "is_lbd_sample", "is_eg8_sample", "is_eg8_fc_sample", "xs_batch", "fc_batch",
                                          "Generation", "alb", "bun", "bw", "cr", "gluc", "igf1", "phos",
                                          "potas", "sodium", "Birth_date", "Draw_date", "death_ES", "death_FD")) %>%
  romic::center_tomic(c("corrected_log2_abundance", "normalized_log2_abundance"))
```

# Consolidate Genetics and Informatics Results

```{r find_wdl_outputs}
featurization_parameters <- featurization_parameterization(
  file.path(params$repo_path, "bioinformatics", "1_wdl_processes", "do_config.json"),
  params$out_base_dir
)

all_normalized_peaks <- featurization_parameters %>%
  mutate(
    peaks_path = file.path(params$project_path, dataset_path, "out_files", "normalized_small_molecules.Rds"),
    peaks_dat = purrr::map(peaks_path, readRDS)
    ) %>%
  dplyr::select(data_type, data_partition, peaks_dat)
```

```{r genetics}
read_physical_map <- function (ls_data_dir) {
  
  # copied from Kevin's measure_qtls.R
  
  pmap_file <- file.path(ls_data_dir, "mm_uwisc_v1_filter_200325.pmap.csv")
  genoprob_file <- file.path(ls_data_dir, "prob.8state_jax_shock_191218.Rdata")
  
  load(genoprob_file)
  
  physical_map <- qtl2::read_csv(pmap_file, rownames_included = FALSE)
  
  lt <- list()
  for(n in names(apr)){
      marker <-(dimnames(apr[[n]])[[3]])
      lt[[n]] <- as.data.frame(marker)
      #print(paste(n, length(tmp)))
  }
  markdf <- do.call(plyr::rbind.fill, lt)  #pylr function to collapse list of dfs to sinlge column
  physical_map_edit <- inner_join(physical_map, markdf, by = c("marker"))
  
  print('dim(physical_map_edit)')
  print(dim(physical_map_edit))
  
  # remotes::install_github("rqtl/qtl2convert")
  
  # Convert map data frame to list
  pmap <- physical_map_edit %>% 
    # Make sure the chromosoms are in the factored properly, if we don't do this
    data.frame() %>% qtl2convert::map_df_to_list(marker_column = 'marker' , pos_column = 'pos', chr_column= 'chr')
  
  #convert pmap value from char to numeric
  count_num_markers = 0
  for(n in names(pmap)){
    #use list apply, lapply, to convert char to numeric. this is imp to use to keep markers as row names 
    pmap[[n]] <- lapply(pmap[[n]], function(x) as.numeric(x))
    #transform to mega_bases - make it compatibile with snp_scan --- cc_variants.sqlite data
    #pmap[[n]] <- lapply(pmap[[n]], function(x) x /1000000)  
    pmap[[n]] <- unlist(pmap[[n]])
    count_num_markers = count_num_markers + length( pmap[[n]] )          
  }
  
  return(pmap)
}

process_trait <- function (feature_name, measure, qtl_results_path) {
  
  feature_dir <- file.path(qtl_results_path, feature_name)
  
  lod_path <- file.path(file.path(feature_dir, paste0(measure, ".lod.Rdata")))
  null_lod_paths <- file.path(feature_dir, paste0(measure, ".permute100.Rdata"))
  
  if (!file.exists(lod_path) | !file.exists(null_lod_paths)) {
    return(NULL)
  }
  
  load(lod_path)
  # 75K SNP scan (Rqtl2 object)
  load(null_lod_paths)
  
  out <- tibble(
    snp_scan_data = list(snp_scan_data),
    permute_scan = list(permute_scan)
  )
  
  return(out)
}

qtl_results_path <- file.path(params$project_path, "genetics", "qtl_mapping") 
pmap <- read_physical_map(file.path(params$project_path, "genetics", "pmap_files"))

# identify all traits w/ qtl results
molecules <- list.files(qtl_results_path)
measures <- c("average_abundance", "early_age", "late_age")

all_qtl_results <- tidyr::crossing(
  feature_name = molecules,
  measure = measures
) %>%
  dplyr::mutate(trait_results = purrr::map2(
    feature_name,
    measure,
    process_trait,
    qtl_results_path = qtl_results_path
    )) %>%
  # drop null traits (these are just non-abundance traits which were not
  # rerun when updating the qtl pipeline)
  filter(!purrr::map_lgl(trait_results, is.null)) %>%
  tidyr::unnest(trait_results)
```

# Export to GCS

```{r upload_gcp_assets}
shiny_assets_dir <- file.path(params$project_path, "doomics_public_gcs_data")
if (!dir.exists(shiny_assets_dir)) {
  dir.create(shiny_assets_dir)
}

# data modality-specific summaries which include additional metadata
saveRDS(feature_design_list, file.path(shiny_assets_dir, "feature_design_list.Rds"))
# statistical significance of select regression terms
saveRDS(model_signif, file.path(shiny_assets_dir, "model_signif.Rds"))
# summaries of individual models fit with parameters, fits, AICc, ...
# for model comparison and pathway delta age
saveRDS(lm_models, file.path(shiny_assets_dir, "lm_models.Rds"))
# statistical significance for all regression terms in all models
file.copy(
  file.path(diffex_outdir, "out_files", "all_model_signif.Rds"),
  file.path(shiny_assets_dir, "all_model_signif.Rds"),
  overwrite = TRUE
)

# features_w_design - all features and samples with lots of metadata
projected_features_with_design_aug <- projected_features_with_design %>%
  dplyr::left_join(feature_clusters, by = "feature_name") %>%
  dplyr::mutate(
    cluster = ifelse(is.na(cluster), -1, cluster),
    cluster = factor(cluster, levels = sort(unique(cluster)))
    )

saveRDS(
  projected_features_with_design_aug,
  file.path(shiny_assets_dir, "features_with_design.Rds")
  )

projected_features_with_design_aug %>%
  filter(data_modality %in% c("metabolomics", "lipidomics", "proteomics")) %>%
  saveRDS(file.path(shiny_assets_dir, "features_with_design_core.Rds"))

# features_w_design as a T\*Omic object
saveRDS(tidy_features, file.path(shiny_assets_dir, "tidy_features.Rds"))

tidy_features %>%
    romic::filter_tomic(
      filter_type = "category",
      filter_table = "features",
      filter_variable = "data_type",
      c("lipids-neg", "lipids-pos", "metabolites-neg", "metabolites-pos", "tryptic-compms")
    ) %>%
  saveRDS(file.path(shiny_assets_dir, "tidy_features_core.Rds"))

tidy_features %>%
    romic::filter_tomic(
      filter_type = "category",
      filter_table = "features",
      filter_variable = "data_type",
      c("peptides")
    ) %>%
  saveRDS(file.path(shiny_assets_dir, "tidy_features_peptides.Rds"))

# genetics and bioinformatics summaries

saveRDS(all_normalized_peaks, file.path(shiny_assets_dir, "all_normalized_peaks.Rds"))
saveRDS(all_qtl_results, file.path(shiny_assets_dir, "all_qtl_results.Rds"))

# move specific informatics outputs

informatics_assets_dir <- file.path(shiny_assets_dir, "select_informatics_plots")
if (!dir.exists(informatics_assets_dir)) {
  dir.create(informatics_assets_dir)
}

# metabolomic & lipidomics plots

injection_plots_df <- featurization_parameters %>%
  filter(set == "1/2", data_type %in% c("metabolites-neg", "metabolites-pos")) %>%
  mutate(
    plot_path = file.path(params$project_path, dataset_path, "out_figures", "injection_comparison-1.png"),
    exists = file.exists(plot_path),
    updated_name = glue::glue("{data_type}_injection_comparison.png"),
    updated_path = file.path(shiny_assets_dir, "select_informatics_plots", updated_name)
    ) 

# move and rename figures
purrr::walk2(
  injection_plots_df$plot_path,
  injection_plots_df$updated_path,
  file.copy
)

# proteomics plots

select_proteomics_plots <- c(
  "batch_post_correction-1.png",
  "tmt_ra_correlations-1.png",
  "tmt_bridge_correlations-1.png"
  )

proteomics_informatics_plot_path <- file.path(params$project_path, params$out_base_dir, "proteomics", "out_figures")

purrr::walk2(
  file.path(proteomics_informatics_plot_path, select_proteomics_plots),
  file.path(shiny_assets_dir, "select_informatics_plots", select_proteomics_plots),
  file.copy
)

# add to GCS
# gcloud config set project <<PROJECT_ID>>
# gsutil cp -r <<PATH_TO_PROJECT_DATA>>/doomics_public_gcs_data/* gs://<<BUCKET_ID>>/
```
